BAU: Add httpEgressSafelist

adityapahuja opened the pull request:
There is an application which wants to talk to another service on the Internet over HTTP. httpsEgressSafelist currently only whitelists services over https. This change adds httpEgressSafelist to allow the application to talk to the service on the Internet over HTTP instead of HTTPS.

Author: @adityapahuja

adityapahuja committed:
Add httpEgressSafelist

There is an application which wants to talk to another service on the Internet over HTTP. httpsEgressSafelist currently only whitelists services over https. This change adds httpEgressSafelist to allow the application to talk to the service e.g. OCSP on the Internet over HTTP.

Author: @adityapahuja

issyl0 commented:
You'll need to add your GPG key to your GitHub account for GitHub (and GSP) to count this as a valid PR and commit.

adityapahuja commented:
I have added GPG key to my GitHub account.

adityapahuja commented:
The commit has "Verified". Looks like a valid commit.

blairboy362 commented on charts/gsp-cluster/templates/02-gsp-system/egress-safelist.yaml:
```diff
@@ -0,0 +1,78 @@
+  ports:
+  - name: {{ $egressFqdn.name }}
+    number: {{ $egressFqdn.portNumber }}
+    protocol: {{ $egressFqdn.protocol }}
```
I don't think this will work how you intend it do. You'll need to make this suite of templates protocol specific, as the structure of `VirtualService` and `DestinationRule` changes depending on whether it's HTTP, HTTPS, TLS or just TCP. So I would make this an "expose HTTP egress safelist" and set it up for that.

adityapahuja replied:
I have reverted the changes back to the original change.

philandstuff commented on charts/gsp-cluster/templates/02-gsp-system/http-egress-safelist.yaml:
```diff
@@ -0,0 +1,71 @@
+  - port:
+      number: 80
+      name: tls-https-egress-safelist-{{ $egressFqdn.name }}
+      protocol: HTTP
```
can this still be istio mTLS as far as the egressgateway? (I don't pretend to understand this stuff)

adityapahuja replied:
I have renamed it to http-egress instead of tls-https-egress.

blairboy362 replied:
I think @philandstuff is correct; some things need changing here for mTLS to the egressgateway. I'll try to put the suggestions in and we'll see how we go.

blairboy362 replied:
```suggestion
      protocol: HTTPS
```

adityapahuja replied:
This has been done.

philandstuff commented:
I'd really like @samcrang to review this one - I don't understand all of this stuff.

I think in an ideal world, we'd have egress http be tunnelled through mTLS as far as the egressgateway and then HTTP over public internet, but it doesn't matter too much.

Krenair reviewed (dismissed):
`diff charts/gsp-cluster/templates/02-gsp-system/https-egress-safelist.yaml charts/gsp-cluster/templates/02-gsp-system/http-egress-safelist.yaml` looks fine

blairboy362 commented on charts/gsp-cluster/templates/02-gsp-system/http-egress-safelist.yaml:
```diff
@@ -0,0 +1,71 @@
+  servers:
+  - port:
+      number: 80
+      name: http-egress-safelist-{{ $egressFqdn.name }}
```
```suggestion
      name: https-http-egress-safelist-{{ $egressFqdn.name }}
```

adityapahuja replied:
This has been done.

blairboy362 commented on charts/gsp-cluster/templates/02-gsp-system/http-egress-safelist.yaml:
```diff
@@ -0,0 +1,71 @@
+      protocol: HTTP
+    hosts:
+    - {{ $egressFqdn.fqdn}}
+---
```
```suggestion
    tls:
      mode: MUTUAL
      serverCertificate: /etc/certs/cert-chain.pem
      privateKey: /etc/certs/key.pem
      caCertificates: /etc/certs/root-cert.pem
---
```

adityapahuja replied:
This has been done.

blairboy362 commented on charts/gsp-cluster/templates/02-gsp-system/http-egress-safelist.yaml:
```diff
@@ -0,0 +1,71 @@
+  - {{ $egressFqdn.fqdn}}
+  gateways:
+  - {{ $.Release.Name }}-http-egress-safelist-{{ $egressFqdn.name }}
+  tcp:
```
```suggestion
  http:
```

adityapahuja replied:
This has been done.

blairboy362 reviewed (changes requested):
I've suggested the changes...

If it's easier to follow, see the examples given [here](https://istio.io/docs/tasks/traffic-management/egress/egress-gateway/#egress-gateway-for-http-traffic)

adityapahuja commented:
@blairboy362 : Looking at the example, do we need to add DestinationRule and mesh to a list of gateways in VirtualService?

adityapahuja committed:
Update an egress gateway to use mutual TLS enabled

This change updates an egress gateway to use mutual TLS enabled so that traffic coming from applications go to the egress gateway securely.

Co-authored-by: Daniel Blair <daniel.blair@digital.cabinet-office.gov.uk>
